src/rate.rs

```diff
@@ -0,0 +1,210 @@
+        self.integer
+    }
+
+    pub fn inverse_decimal_exponent(&self) -> usize {
+        self.inv_dec_exp
+    }
+
+    pub fn new(integer: u64, inverse_decimal_exponent: usize) -> anyhow::Result<Self> {
+        let (integer, inv_dec_exp) = Rate::reduce(integer, inverse_decimal_exponent);
+
+        let integer =
+            u32::try_from(integer).map_err(|_| anyhow::anyhow!("Value is unexpectedly large."))?;
+        Ok(Rate {
+            integer,
+            inv_dec_exp,
+        })
+    }
+
+    pub fn from_f64(rate: f64) -> anyhow::Result<Rate> {
+        if rate.is_sign_negative() {
+            anyhow::bail!("Spread must be positive");
+        }
+
+        if !rate.is_finite() {
+            anyhow::bail!("Spread must be finite")
+        }
+
+        let mut rate = rate.to_string();
+        let decimal_index = rate.find('.');
+        match decimal_index {
+            None => {
+                let rate = u64::from_str(&rate)
+                    .map_err(|_| anyhow::anyhow!("Rate is unexpectedly large."))?;
+                Rate::new(rate, 0)
+            }
+            Some(decimal_index) => {
+                let mantissa = rate.split_off(decimal_index + 1);
+                if mantissa.len() > Self::MAX_PRECISION {
+                    anyhow::bail!("Precision of the rate is too high.")
+                } else {
+                    // Removes the trailing decimal point
+                    rate.truncate(rate.len() - 1);
+                    let integer = rate;
+
+                    let inv_dec_exp = mantissa.len();
+
+                    let integer = u64::from_str(&format!("{}{}", integer, mantissa))
+                        .map_err(|_| anyhow::anyhow!("Rate is unexpectedly large"))?;
+                    Rate::new(integer, inv_dec_exp)
+                }
+            }
+        }
+    }
+
+    /// If the integer part ends with 0 and the inverse_decimal_exponent is not null
+    /// then we can reduce the representation by removing zero and decrementing the inverse
+    /// exponent. For example:
+    /// Rate { integer: 1000, inv_dec_exp: 1 } becomes
+    /// Rate { integer: 100, inv_dec_exp: 0 }.
+    fn reduce(integer: u64, inv_dec_exp: usize) -> (u64, usize) {
+        let mut integer_str = integer.to_string();
+        if integer_str.len() > 1 && integer_str.ends_with('0') && inv_dec_exp != 0 {
+            integer_str.truncate(integer_str.len() - 1);
+            let inv_dec_exp = inv_dec_exp - 1;
+            let integer = u64::from_str(&integer_str).expect("an integer");
+            Rate::reduce(integer, inv_dec_exp)
+        } else {
+            (integer, inv_dec_exp)
+        }
+    }
+}
+
+/// Spread: percentage to be added on top of a rate or amount with
+/// a maximum precision of 2 decimals
+// 0 is the rate * 100.
+#[derive(Clone, Copy, Debug)]
+pub struct Spread(u32);
+
+impl Spread {
+    /// Input is the spread in percent: 5.0 is 5%
+    pub fn new(spread: f64) -> anyhow::Result<Spread> {
```

Comment:
@thomaseizinger what do you mean by:

> force the user to already give us the number as a "cent", i.e. 5 and accept a u8

The spread is a percentage, it does not have know about "cents", or did I misunderstand something?
Are you thinking about adding a fixed spread? (as in a fixed value rather than a percentage)

Reply:
As discussed during the call, accepting a `&str` here should be fine but `f64` is fine too. `f64` is useful to decode the response from Kraken.

I like the `u16` idea.

Reply:
However there are no available conversion from float primitives to integer ones.

Reply:
> As discussed during the call, accepting a `&str` here should be fine but `f64` is fine too. `f64` is useful to decode the response from Kraken.
> 
> I like the `u16` idea.

To get good benefits out of a strongly typed language like Rust, it is important to encode as much information as possible in the type system. That means, if the response from Kraken is a float, decoding that into an f64 is fine. But the very next step should be a conversion into a datatype that is best suited for the core of the program so the core doesn't have to make assertions or assumptions about the data.

In our case, the response from Kraken are USD cents which fit perfectly into a u32. Hence, I'd suggest that the output of the Kraken module is already a u32. That frees this core logic from a lot of error cases around floats (negative numbers, infinite, too high precision, etc).

Reply:
getting the rate from kraken to using it here will be addressed in a follow up PR